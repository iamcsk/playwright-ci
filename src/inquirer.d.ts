declare module '@inquirer/confirm' {
  type ConfirmArgs = {
    default?: boolean;
    message: string;
  };

  const confirm: (args: ConfirmArgs) => Promise<boolean>;
  export default confirm;
}

declare module '@inquirer/select' {
  type SelectArgs = {
    choices: Array<{ name: string; value: string | null }>;
    message: string;
  };

  const select: (args: SelectArgs) => Promise<string | null>;
  export default select;
}
