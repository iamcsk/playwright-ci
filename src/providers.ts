export const CI_PROVIDERS: Array<{
  label: string;
  name: string | null;
  path: string;
  qawolfPath?: string;
}> = [
  { label: 'Azure DevOps', name: 'azure', path: 'azure-pipelines.yml' },
  {
    label: 'Bitbucket Pipelines',
    name: 'bitbucket',
    path: 'bitbucket-pipelines.yml',
  },
  { label: 'CircleCI', name: 'circleci', path: '.circleci/config.yml' },
  {
    label: 'GitHub Actions',
    name: 'github',
    path: '.github/workflows/playwright.yml',
    qawolfPath: '.github/workflows/qawolf.yml',
  },
  { label: 'GitLab CI/CD', name: 'gitlab', path: '.gitlab-ci.yml' },
  { label: 'Jenkins', name: 'jenkins', path: 'Jenkinsfile' },
  { label: 'Skip CI setup', name: null, path: '' },
];
