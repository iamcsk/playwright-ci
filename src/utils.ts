import { promises as fs } from 'fs';

// https://stackoverflow.com/a/57708635
export const pathExists = async (path: string): Promise<boolean> =>
  !!(await fs.stat(path).catch(() => false));
