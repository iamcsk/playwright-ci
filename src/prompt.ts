import confirm from '@inquirer/confirm';
import select from '@inquirer/select';
import { pathExists } from 'fs-extra';
import { CI_PROVIDERS } from './providers';

export const promptConfirmOverwrite = (path: string): Promise<boolean> =>
  confirm({
    message: `"${path}" already exists, overwrite it?`,
    default: false,
  });

export const promptOverwrite = async (path: string): Promise<boolean> => {
  const exists = await pathExists(path);
  if (!exists) return true;

  return promptConfirmOverwrite(path);
};

export const promptCiProvider = async (): Promise<string | null> => {
  return select({
    message: 'Select a CI provider',
    choices: CI_PROVIDERS.map(provider => {
      return { name: provider.label, value: provider.name };
    }),
  });
};
