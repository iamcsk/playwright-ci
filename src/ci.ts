import { promises as fs, readFileSync } from 'fs';
import { compile } from 'handlebars';
import { join, resolve } from 'path';
import { promptCiProvider } from './prompt';
import { promptOverwrite } from './prompt';
import { CI_PROVIDERS } from './providers';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { version } = require('../package');

interface SaveCiTemplateArgs {
  provider: string;
  qawolf: boolean;
}

export const buildCiTemplate = ({
  provider,
  qawolf,
}: SaveCiTemplateArgs): string => {
  const templateFn = compile(
    readFileSync(resolve(__dirname, `../static/${provider}.hbs`), 'utf8'),
  );

  return templateFn({ qawolf, version });
};

export const saveCiTemplate = async ({
  provider,
  qawolf,
}: SaveCiTemplateArgs): Promise<void> => {
  const fullProvider = CI_PROVIDERS.find(p => p.name === provider);
  if (!fullProvider) {
    throw new Error(`No template for CI provider ${provider}`);
  }

  const providerPath = qawolf
    ? fullProvider.qawolfPath || fullProvider.path
    : fullProvider.path;

  const outputPath = join(process.cwd(), providerPath);

  const shouldOverwrite = await promptOverwrite(outputPath);
  if (!shouldOverwrite) return;

  const template = buildCiTemplate({ provider, qawolf });
  await fs.writeFile(outputPath, template, 'utf8');

  console.log(`Saved ${provider} template to ${outputPath}`);
};

export const install = async (qawolf = false): Promise<void> => {
  const provider = await promptCiProvider();
  if (!provider) return;

  await saveCiTemplate({ provider, qawolf });
};
