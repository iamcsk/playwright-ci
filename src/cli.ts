import select from '@inquirer/select';
import { CiProvider, saveCiTemplate } from './ci';

export const CI_PROVIDERS: Array<{ name: string; value: CiProvider | null }> = [
  { name: 'Azure DevOps', value: 'azure' },
  { name: 'Bitbucket Pipelines', value: 'bitbucket' },
  { name: 'CircleCI', value: 'circleci' },
  { name: 'GitHub Actions', value: 'github' },
  { name: 'GitLab CI/CD', value: 'gitlab' },
  { name: 'Jenkins', value: 'jenkins' },
  { name: 'Skip CI setup', value: null },
];

const getCiProvider = async (): Promise<CiProvider | null> => {
  return select({
    message: 'Select a CI provider',
    choices: CI_PROVIDERS,
  });
};

export const install = async (qawolf: boolean = false): Promise<void> => {
  const provider = await getCiProvider();
  if (!provider) return;

  await saveCiTemplate({ provider, qawolf });
};
