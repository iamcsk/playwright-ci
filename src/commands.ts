import * as program from 'commander';

import { saveCiTemplate } from './ci';

export type CiProvider =
  | 'azure'
  | 'bitbucket'
  | 'circleci'
  | 'github'
  | 'gitlab'
  | 'jenkins';

interface AddCiCommandsArgs {
  program: program.CommanderStatic;
  qawolf: boolean;
}

export const COMMANDS: { command: CiProvider; label: string }[] = [
  { command: 'azure', label: 'an Azure Pipeline' },
  { command: 'bitbucket', label: 'a Bitbucket Pipeline' },
  { command: 'circleci', label: 'CircleCI' },
  { command: 'github', label: 'a GitHub Action' },
  { command: 'gitlab', label: 'Gitlab CI/CD' },
  { command: 'jenkins', label: 'Jenkins' },
];

export const addCiCommands = ({ program, qawolf }: AddCiCommandsArgs): void => {
  COMMANDS.forEach(({ command, label }) => {
    program
      .command(command)
      .description(`set up ${label}`)
      .action(async () => {
        await saveCiTemplate({ provider: command, qawolf });
      });
  });
};
