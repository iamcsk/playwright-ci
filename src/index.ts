#!/usr/bin/env node
const isCLI = !module.parent;
if (isCLI) {
  require('./cli');
}

export { addCiCommands } from './commands';
