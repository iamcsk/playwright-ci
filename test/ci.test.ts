import * as ci from '../src/ci';
import { CI_PROVIDERS } from '../src/providers';

const { buildCiTemplate } = ci;

describe('buildCiTemplate', () => {
  it('builds playwright templates', () => {
    CI_PROVIDERS.forEach(item => {
      if (!item.name) return;
      const provider = item.name;
      expect(buildCiTemplate({ provider, qawolf: false })).toMatchSnapshot(
        provider,
      );
    });
  });

  it('builds qawolf templates', () => {
    CI_PROVIDERS.forEach(item => {
      if (!item.name) return;

      const provider = item.name;
      expect(buildCiTemplate({ provider, qawolf: true })).toMatchSnapshot(
        `${provider}_qawolf`,
      );
    });
  });
});
