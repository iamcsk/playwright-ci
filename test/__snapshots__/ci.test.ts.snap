// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCiTemplate builds playwright templates: azure 1`] = `
"# configure the pipeline to run based on different triggers
# see https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema#triggers
pool:
  vmImage: \\"ubuntu-18.04\\"

steps:
  - task: NodeTool@0
    inputs:
      versionSpec: \\"12.x\\"

  - script: |
      sudo apt update
      # chromium dependencies
      sudo apt-get install libgbm1
      # webkit dependencies
      sudo apt-get install libwoff1 libopus0 libwebp6 libwebpdemux2 libenchant1c2a libgudev-1.0-0 libsecret-1-0 libhyphen0 libgdk-pixbuf2.0-0 libegl1 libgles2 libevent-2.1-6 libnotify4 libvpx5 libxslt1.1
    displayName: \\"Install browser dependencies\\"

  - script: npm install

  # # Start local server
  # - script: npm run start & npx wait-on http://localhost:3000

  # replace below with command you want to run, example for running a script below
  # - script: node myScript.js
  - script: npm test
"
`;

exports[`buildCiTemplate builds playwright templates: bitbucket 1`] = `
"pipelines:
  default:
    - step:
        name: Playwright Tests
        image: qawolf/playwright-ci:v0.10.0
        caches:
          - node
        script:
          - npm install
        
          ## Start local server
          # - npm run start & npx wait-on http://localhost:3000
          # replace below with command you want to run, example for running a script below
          # - node myScript.js
          - npm test
"
`;

exports[`buildCiTemplate builds playwright templates: circleci 1`] = `
"version: 2
jobs:
  build:
    docker:
      - image: qawolf/playwright-ci:v0.10.0
    steps:
      - checkout

      - restore_cache:
          key: dependency-cache-{{ checksum \\"package-lock.json\\" }}

      - run:
          command: npm install

      - save_cache:
          key: dependency-cache-{{ checksum \\"package-lock.json\\" }}
          paths:
            - /root/cache
            - ./node_modules

      - run:
          command: |
            # # Start local server
            # npm run start & 
            # npx wait-on http://localhost:3000
            # replace below with command you want to run, example for running a script below
            # node myScript.js
            npm test

# # example for running on a schedule, edit to suit your use case
# # documentation: https://circleci.com/docs/2.0/configuration-reference/#schedule
# workflows:
#   version: 2
#   on_schedule:
#     jobs:
#       - build
#     triggers:
#       - schedule:
#           # test on schedule using cron syntax
#           cron: \\"0 * * * *\\" # every hour
#           filters:
#             branches:
#               only:
#                 - master"
`;

exports[`buildCiTemplate builds playwright templates: github 1`] = `
"name: playwright
on:
  push:
    # test every branch
    # edit below if you only want certain branches tested
    branches: \\"*\\"
  # schedule:
  #   # test on schedule using cron syntax
  #   - cron: \\"0 * * * *\\" # every hour
jobs:
  test:
    runs-on: ubuntu-18.04

    steps:
      - uses: actions/checkout@v2

      - uses: actions/setup-node@v1

      - uses: microsoft/playwright-github-action@v1

      - uses: actions/cache@v1
        with:
          path: ~/.npm
          key: \${{ runner.os }}-node-\${{ hashFiles('**/package-lock.json') }}
          restore-keys: |
            \${{ runner.os }}-node-

      - run: npm install

      # - name: Start local server
      #   run: npm run start & npx wait-on http://localhost:3000

      # replace below with command you want to run, example for running a script below
      # - run: node myScript.js
      - run: npm test
"
`;

exports[`buildCiTemplate builds playwright templates: gitlab 1`] = `
"qawolf:
  image: qawolf/playwright-ci:v0.10.0

  cache:
    paths:
      - /root/cache/
      - node_modules/

  script:
    - npm install
    # # Start local server
    # - npm run start & npx wait-on http://localhost:3000
    # replace below with command you want to run, example for running a script below
    # - node myScript.js
    - npm test
"
`;

exports[`buildCiTemplate builds playwright templates: jenkins 1`] = `
"pipeline {
    agent {
        docker {
            image 'qawolf/playwright-ci:v0.10.0'
        }
    }
    stages {
        stage('Build') { 
            steps {
                sh 'npm install'
            }
        }
        stage('Test') {
            steps {
                // // Start local server
                // sh 'npm run start & npx wait-on http://localhost:3000'
                // replace below with command you want to run, example for running a script below
                // sh 'node myScript.js'
                sh 'npm test'
            }
        }
    }
}"
`;

exports[`buildCiTemplate builds qawolf templates: azure_qawolf 1`] = `
"# configure the pipeline to run based on different triggers
# see https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=azure-devops&tabs=schema#triggers
pool:
  vmImage: \\"ubuntu-18.04\\"

steps:
  - task: NodeTool@0
    inputs:
      versionSpec: \\"12.x\\"

  - script: |
      sudo apt update
      # chromium dependencies
      sudo apt-get install libgbm1
      # webkit dependencies
      sudo apt-get install libwoff1 libopus0 libwebp6 libwebpdemux2 libenchant1c2a libgudev-1.0-0 libsecret-1-0 libhyphen0 libgdk-pixbuf2.0-0 libegl1 libgles2 libevent-2.1-6 libnotify4 libvpx5 libxslt1.1
      # ffmpeg
      sudo apt-get install -y ffmpeg
    displayName: \\"Install browser dependencies\\"

  - script: npm install

  # # Start local server
  # - script: npm run start & npx wait-on http://localhost:3000

  - script: npx qawolf test --headless
    env:
      # configure tests with environment variables
      FFMPEG_PATH: /usr/bin/ffmpeg # for recording video
      QAW_ARTIFACT_PATH: $(System.DefaultWorkingDirectory)/artifacts

  - publish: $(System.DefaultWorkingDirectory)/artifacts
    artifact: artifacts
    condition: always()
"
`;

exports[`buildCiTemplate builds qawolf templates: bitbucket_qawolf 1`] = `
"pipelines:
  default:
    - step:
        name: QA Wolf Tests
        image: qawolf/playwright-ci:v0.10.0
        caches:
          - node
        script:
          - npm install
        
          ## Start local server
          # - npm run start & npx wait-on http://localhost:3000
          - QAW_ARTIFACT_PATH=artifacts npx qawolf test --headless
        artifacts:
          - artifacts/**
"
`;

exports[`buildCiTemplate builds qawolf templates: circleci_qawolf 1`] = `
"version: 2
jobs:
  build:
    docker:
      - image: qawolf/playwright-ci:v0.10.0
    steps:
      - checkout

      - restore_cache:
          key: dependency-cache-{{ checksum \\"package-lock.json\\" }}

      - run:
          command: npm install

      - save_cache:
          key: dependency-cache-{{ checksum \\"package-lock.json\\" }}
          paths:
            - /root/cache
            - ./node_modules

      - run:
          command: |
            # # Start local server
            # npm run start & 
            # npx wait-on http://localhost:3000
            npx qawolf test --headless
          environment:
            # configure tests with environment variables
            QAW_ARTIFACT_PATH: /tmp/artifacts
      - store_artifacts:
          path: /tmp/artifacts

# # example for running on a schedule, edit to suit your use case
# # documentation: https://circleci.com/docs/2.0/configuration-reference/#schedule
# workflows:
#   version: 2
#   on_schedule:
#     jobs:
#       - build
#     triggers:
#       - schedule:
#           # test on schedule using cron syntax
#           cron: \\"0 * * * *\\" # every hour
#           filters:
#             branches:
#               only:
#                 - master"
`;

exports[`buildCiTemplate builds qawolf templates: github_qawolf 1`] = `
"name: qawolf
on:
  push:
    # test every branch
    # edit below if you only want certain branches tested
    branches: \\"*\\"
  # schedule:
  #   # test on schedule using cron syntax
  #   - cron: \\"0 * * * *\\" # every hour
jobs:
  test:
    runs-on: ubuntu-18.04

    steps:
      - uses: actions/checkout@v2

      - uses: actions/setup-node@v1

      - uses: microsoft/playwright-github-action@v1

      - uses: actions/cache@v1
        with:
          path: ~/.npm
          key: \${{ runner.os }}-node-\${{ hashFiles('**/package-lock.json') }}
          restore-keys: |
            \${{ runner.os }}-node-

      - run: npm install

      # - name: Start local server
      #   run: npm run start & npx wait-on http://localhost:3000

      - run: npx qawolf test --headless
        env:
          # configure tests with environment variables
          FFMPEG_PATH: /usr/bin/ffmpeg # for recording video
          QAW_ARTIFACT_PATH: \${{ github.workspace }}/artifacts
          # you can also use GitHub secrets for environment variables
          # https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets
          # LOGIN_PASSWORD: \${{ secrets.PASSWORD }}
      
      - name: Upload Artifacts
        if: always()
        uses: actions/upload-artifact@master
        with:
          name: qawolf
          path: \${{ github.workspace }}/artifacts
"
`;

exports[`buildCiTemplate builds qawolf templates: gitlab_qawolf 1`] = `
"qawolf:
  image: qawolf/playwright-ci:v0.10.0

  cache:
    paths:
      - /root/cache/
      - node_modules/

  script:
    - npm install
    # # Start local server
    # - npm run start & npx wait-on http://localhost:3000
    - npx qawolf test --headless

  variables:
    # configure tests with environment variables
    QAW_ARTIFACT_PATH: $CI_PROJECT_DIR/artifacts
  
  artifacts:
    when: always
    paths:
      - $CI_PROJECT_DIR/artifacts
    expire_in: 1 week
"
`;

exports[`buildCiTemplate builds qawolf templates: jenkins_qawolf 1`] = `
"pipeline {
    agent {
        docker {
            image 'qawolf/playwright-ci:v0.10.0'
        }
    }
    stages {
        stage('Build') { 
            steps {
                sh 'npm install'
            }
        }
        stage('Test') {
            environment { 
                // configure tests with environment variables
                QAW_ARTIFACT_PATH = './artifacts'
            }
            steps {
                // // Start local server
                // sh 'npm run start & npx wait-on http://localhost:3000'
                sh 'npx qawolf test --headless'
            }
        }
        post {
            always {
                archiveArtifacts(artifacts: 'artifacts/**/*.*', fingerprint: true) 
            }
        }
    }
}"
`;
